### PR TITLE
Feature lp 16 signin be

### PR DESCRIPTION
**Pull request by ShreyasMandlik**
# Checklist

- [ ] My code follows the style guidelines of this project and is properly formatted
- [ ] I have performed a self-review of my own changes
- [ ] All existing tests pass, and I have added new ones if required

### CODE REVIEW

File: backend/src/auth/Dto/userDto.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
...
+
+export class UserDto {
+  @IsEmail()
+  @IsNotEmpty()
+  @Contains('@incubyte.co')
+  email: string;
+
+  @IsNotEmpty()
+  password: string;
+}
```
Comment (utsav00) [inline backend/src/auth/Dto/userDto.ts:11]: Should be `dto` and the file name should follow convention - `user.dto.ts`

File: backend/src/auth/auth.controller.spec.ts [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+
+  it('should be able to create user', async () => {
+    const user: UserDto = {
+      email: 'john@incubyte.co',
+      password: '123',
+    };
+    jest
+      .spyOn(service, 'signup')
+      .mockResolvedValueOnce({
+        ...user,
```
Comment (utsav00) [inline backend/src/auth/auth.controller.spec.ts:42]: Tests should be as dumb as possible. Don't use the spread operator, instead just mention the literals

**Comment by utsav00** [comment]
Add path aliasing for imports

File: backend/src/auth/auth.service.ts [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+@Injectable()
+export class AuthService {
+  constructor(private readonly prismaService: PrismaService, private readonly jwtService : JwtService) {}
+
+  async signup(user: UserDto) {
+    const userbyEmail = await this.prismaService.user.findFirst({
+      where: { email: user.email },
+    });
+
+    if (userbyEmail) {
```
Comment (abdulkalam1233) [inline backend/src/auth/auth.service.ts:17]: Create separate method to check use exist

File: backend/src/auth/auth.service.ts [hunk 4]
```diff
@@ -0,0 +1,62 @@
+import { BadRequestException, Injectable} from '@nestjs/common';
+import { UserDto } from './Dto/userDto';
+import { PrismaService } from './../prisma/prisma.service';
+import * as bcrypt from 'bcrypt';
+import { compareSync } from 'bcrypt';
+import { JwtService } from '@nestjs/jwt';
+
+@Injectable()
+export class AuthService {
+  constructor(private readonly prismaService: PrismaService, private readonly jwtService : JwtService) {}
```
Comment (abdulkalam1233) [inline backend/src/auth/auth.service.ts:10]: Formatting

File: backend/src/auth/auth.service.ts [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+
+  async signup(user: UserDto) {
+    const userbyEmail = await this.prismaService.user.findFirst({
+      where: { email: user.email },
+    });
+
+    if (userbyEmail) {
+      throw new BadRequestException('Email Already exists');
+    }
+
```
Comment (abdulkalam1233) [inline backend/src/auth/auth.service.ts:20]: Remove new lines everywhere where new line doesn't make sense like here

File: backend/src/auth/auth.service.ts [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+import * as bcrypt from 'bcrypt';
+import { compareSync } from 'bcrypt';
+import { JwtService } from '@nestjs/jwt';
+
+@Injectable()
+export class AuthService {
+  constructor(private readonly prismaService: PrismaService, private readonly jwtService : JwtService) {}
+
+  async signup(user: UserDto) {
+    const userbyEmail = await this.prismaService.user.findFirst({
```
Comment (abdulkalam1233) [inline backend/src/auth/auth.service.ts:13]: Name it to use  prismaUser if the variable name "user" is conflicting

File: backend/package.json [hunk 7]
```diff
@@ -21,12 +21,18 @@
   },
   "dependencies": {
     "@nestjs/common": "^9.0.0",
+    "@nestjs/config": "^2.3.1",
     "@nestjs/core": "^9.0.0",
+    "@nestjs/jwt": "^10.0.2",
+    "@nestjs/passport": "^9.0.3",
     "@nestjs/platform-express": "^9.0.0",
     "@prisma/client": "^4.11.0",
+    "@types/bcrypt": "^5.0.0",
```
Comment (abdulkalam1233) [inline backend/package.json:30]: Move to dev dependency

File: backend/test/auth.e2e-spec.ts [hunk 8]
```diff
@@ -18,14 +18,25 @@ describe('AppController (e2e)', () => {
 
   it('/signup (POST) - should return the created user', async () => {
         const user = {
-          email: 'john@example.com',
+          email: 'john@incubyte.co',
           password: '123',
         };
-
         const res = await request(app.getHttpServer())
```
Comment (abdulkalam1233) [inline backend/test/auth.e2e-spec.ts:24]: Do separate assertions

File: backend/src/auth/Dto/userDto.ts [hunk 9]
```diff
@@ -1,9 +1,9 @@
-import { IsEmail, IsNotEmpty, Matches } from 'class-validator';
+import { Contains, IsEmail, IsNotEmpty} from 'class-validator';
 
 export class UserDto {
   @IsEmail()
   @IsNotEmpty()
-  @Matches('^[a-zA-Z0-9+_.-]+@[incubyte.co]+$',"",{"message":"Email Id should be an Incubyte Email-Id. "})
+  @Contains('@incubyte.co')
```
Comment (abdulkalam1233) [inline backend/src/auth/Dto/userDto.ts:6]: Update these validations   @IsEmail({
    host_blacklist: ['incubyte.co'],
  })

File: backend/src/auth/auth.controller.ts [hunk 10]
```diff
@@ -4,12 +4,16 @@ import { AuthService } from './auth.service';
 
 @Controller('')
```
Comment (abdulkalam1233) [inline backend/src/auth/auth.controller.ts:5]: We can add a prefix "/auth" which is reference of AuthModule

File: backend/src/auth/auth.module.ts [hunk 11]
```diff
@@ -1,8 +1,23 @@
 import { Module } from '@nestjs/common';
+import { ConfigService } from '@nestjs/config';
+import { JwtModule } from '@nestjs/jwt';
+import { PassportModule } from '@nestjs/passport';
 import { AuthController } from './auth.controller';
 import { AuthService } from './auth.service';
 
 @Module({
+  imports :[PassportModule,
```
Comment (abdulkalam1233) [inline backend/src/auth/auth.module.ts:9]: formatting

File: backend/src/auth/auth.service.spec.ts [hunk 12]
```diff
@@ -3,10 +3,12 @@ import { PrismaService } from '../prisma/prisma.service';
...
 import { UserDto } from './Dto/userDto';
 import { BadRequestException } from '@nestjs/common';
+import { JwtService } from '@nestjs/jwt';
 
 describe('AuthService', () => {
   let service: AuthService;
   let prismaService:PrismaService;
+  let jwtService :JwtService;
 
   beforeEach(async () => {
```
Comment (abdulkalam1233) [inline backend/src/auth/auth.service.spec.ts:13]: Review beforeAll vs beforeEach

**Comment by abdulkalam1233** [comment]
Let's Add a return type on every function. And configure prettier in your IDE

File: backend/src/auth/Dto/userDto.ts (hunk 1 above)
Comment (ShreyasMandlik) [reply backend/src/auth/Dto/userDto.ts]: Done

File: backend/src/auth/auth.controller.spec.ts (hunk 2 above)
Comment (ShreyasMandlik) [reply backend/src/auth/auth.controller.spec.ts]: Done

File: backend/src/auth/auth.service.ts (hunk 3 above)
Comment (ShreyasMandlik) [reply backend/src/auth/auth.service.ts]: Done

File: backend/src/auth/auth.service.ts (hunk 6 above)
Comment (AshPrajapati) [reply backend/src/auth/auth.service.ts]: done

**Pull request merged**